src/osp/tasks/top_execute.cpp

```diff
@@ -0,0 +1,169 @@
+        auto const tasksToRunBits = lgrn::bit_view(tasksToRun);
+        unsigned int const availableCount = tasksToRunBits.count();
+
+        if (availableCount == 0)
+        {
+            break;
+        }
+
+        // Choose first available task
+        auto const task = TaskId(*tasksToRunBits.ones().begin());
+
+        task_start(tags, tasks, rExec, task);
+
+        TopTask &rTopTask = rTaskData.m_taskData.at(std::size_t(task));
+        dataRefs.resize(rTopTask.m_dataUsed.size());
+        for (std::size_t i = 0; i < rTopTask.m_dataUsed.size(); ++i)
+        {
+            dataRefs.at(i) = topData[rTopTask.m_dataUsed[i]].as_ref();
+        }
+
+        TopTaskStatus const status = rTopTask.m_func(context, dataRefs);
+
+        BitSpan_t enqueue = (status == TopTaskStatus::Success) ? tmpEnqueue : BitSpan_t{};
+
+        task_finish(tags, tasks, rExec, task, enqueue);
+    }
+}
+
+void top_enqueue_quick(Tags const& tags, Tasks const& tasks, ExecutionContext& rExec, ArrayView<TagId const> tagsEnq)
+{
+    rExec.m_tagIncompleteCounts  .resize(tags.m_tags.capacity(), 0);
+    rExec.m_tagRunningCounts     .resize(tags.m_tags.capacity(), 0);
+    rExec.m_taskQueuedCounts     .resize(tasks.m_tasks.capacity(), 0);
+
+    std::vector<uint64_t> tagsToRun(tags.m_tags.vec().size());
+    to_bitspan(tagsEnq, tagsToRun);
+    task_enqueue(tags, tasks, rExec, tagsToRun);
+}
+
+static void check_task_dependencies(Tags const& tags, Tasks const& tasks, TopTaskDataVec_t const& taskData, std::vector<uint32_t> &rPath, uint32_t const task, bool& rGood)
+{
+    rPath.push_back(task);
+
+    auto const taskTags2d   = task_tags_2d(tags, tasks);
+    auto const tagDepends2d = tag_depends_2d(tags);
+
+    std::vector<uint64_t> dependencyTagInts(tags.m_tags.vec().size(), 0x0);
+    auto dependencyTagsBits = lgrn::bit_view(dependencyTagInts);
+
+    // Loop through each of this task's tags to get dependent tags
+    auto const taskTagsInts = taskTags2d[task].asContiguous();
+    auto const view = lgrn::bit_view(taskTagsInts);
+    for (uint32_t const currTag : view.ones())
+    {
+        auto const currTagDepends = tagDepends2d[currTag].asContiguous();
+
+        for (TagId const dependTag : currTagDepends)
+        {
+            if (dependTag == lgrn::id_null<TagId>())
+            {
+                break;
+            }
+
+            dependencyTagsBits.set(std::size_t(dependTag));
+        }
+    }
+
+    // Get all tasks that contain any of the dependent tags
+    for (uint32_t const currTask : tasks.m_tasks.bitview().zeros())
+    {
+        if (any_bits_match(taskTags2d[currTask].asContiguous(), dependencyTagInts))
+        {
+            auto const &last = std::end(rPath);
+            if (last == std::find(std::begin(rPath), last, currTask))
+            {
+                check_task_dependencies(tags, tasks, taskData, rPath, currTask, rGood);
+            }
+            else
+            {
+                rGood = false;
```

Comment:
this is the only place where this reference parameter is modified.

i recommend making it a return value instead of a reference param.

Reply:
In this case, I find it easier to understand when there is only one instance of the 'good' boolean.